### PR TITLE
DNMY: updates for JuMP 0.21

### PR DESCRIPTION
**Pull request by lkapelevich**

### CODE REVIEW

**Comment by chriscoey** [comment]
thanks! would be great to have this asap

File: Project.toml [hunk 1]
```diff
@@ -17,8 +17,6 @@ SemialgebraicSets = "8e049039-38e8-557d-ae3a-bc521ccf6204"
 SparseArrays = "2f01184e-e22b-5df5-ae63-d93ebab69eaf"
 
 [compat]
-DynamicPolynomials = "0.3.6"
```
Comment (blegat) [inline Project.toml]: You should keep that

File: Project.toml [hunk 2]
```diff
@@ -17,8 +17,6 @@ SemialgebraicSets = "8e049039-38e8-557d-ae3a-bc521ccf6204"
...
-JuMP = "0.20"
```
Comment (blegat) [inline Project.toml]: Just replace 20 by 21, don't remove it

File: src/variable.jl [hunk 3]
```diff
@@ -30,14 +30,16 @@ matrix_cone_type(::SDSOSPoly) = ScaledDiagonallyDominantConeTriangle
 const PosPoly{PB} = Union{DSOSPoly{PB}, SDSOSPoly{PB}, SOSPoly{PB}}
 
 JuMP.variable_type(m::JuMP.Model, p::PosPoly) = PolyJuMP.polytype(m, p, p.polynomial_basis)
-function PolyJuMP.polytype(m::JuMP.Model, cone::PosPoly,
-                           basis::AbstractPolynomialBasis)
-    return GramMatrix{JuMP.VariableRef, typeof(basis), JuMP.AffExpr}
-end
+
+# TODO can all commented code be deleted?
+# function PolyJuMP.polytype(m::JuMP.Model, cone::PosPoly,
```
Comment (blegat) [inline src/variable.jl]: Yes, delete it :)

File: src/variable.jl [hunk 4]
```diff
@@ -30,14 +30,16 @@ matrix_cone_type(::SDSOSPoly) = ScaledDiagonallyDominantConeTriangle
 const PosPoly{PB} = Union{DSOSPoly{PB}, SDSOSPoly{PB}, SOSPoly{PB}}
 
 JuMP.variable_type(m::JuMP.Model, p::PosPoly) = PolyJuMP.polytype(m, p, p.polynomial_basis)
```
Comment (blegat) [inline src/variable.jl]: You can remove this too, see https://github.com/JuliaOpt/JuMP.jl/issues/2174

File: src/variable.jl [hunk 5]
```diff
@@ -30,14 +30,16 @@ matrix_cone_type(::SDSOSPoly) = ScaledDiagonallyDominantConeTriangle
...
+# TODO can all commented code be deleted?
+# function PolyJuMP.polytype(m::JuMP.Model, cone::PosPoly,
+#                            basis::AbstractPolynomialBasis)
+#     return GramMatrix{JuMP.VariableRef, typeof(basis), JuMP.AffExpr}
+# end
 
 # Sum-of-Squares polynomial
 
-_polytype(m::JuMP.Model, ::PosPoly, x::MVT) where {MT<:MP.AbstractMonomial, MVT<:AbstractVector{MT}} = GramMatrix{JuMP.VariableRef, MT, MVT}
+# _polytype(m::JuMP.Model, ::PosPoly, x::MVT) where {MT<:MP.AbstractMonomial, MVT<:AbstractVector{MT}} = GramMatrix{JuMP.VariableRef, MT, MVT}
```
Comment (blegat) [inline src/variable.jl]: remove

File: test/solvers.jl [hunk 6]
```diff
@@ -13,22 +13,17 @@ mos = try_import(:MosekTools)
...
 # Currently, need to create a file param.csdp to avoid printing, see https://github.com/JuliaOpt/CSDP.jl/issues/15
-csd && push!(sdp_factories, with_opt(CSDP, printlevel=0))
-iscsdp(factory) = occursin("CSDP", string(factory.constructor))
+# csd && push!(sdp_factories, optimizer_with_attributes(CSDP.Optimizer, "printlevel" => 0))
+iscsdp(factory) = occursin("CSDP", string(factory.optimizer_constructor))
 # Need 54000 iterations for sosdemo3 to pass on Linux 64 bits
 # With 55000, sosdemo3 passes for every platform except Windows 64 bits on AppVeyor
-scs && push!(sdp_factories, with_opt(SCS, eps=1e-6, max_iters=60000, verbose=0))
-isscs(factory) = occursin("SCS", string(factory.constructor))
+# scs && push!(sdp_factories, optimizer_with_attributes(SCS.Optimizer, "eps" => 1e-6, "max_iters" => 60000, "verbose" => 0))
```
Comment (blegat) [inline test/solvers.jl]: Why is this commented ?

File: test/solvers.jl [hunk 7]
```diff
@@ -13,22 +13,17 @@ mos = try_import(:MosekTools)
...
 # Need at least Mosek 8 for sosdemo3 to pass; see:
 # https://github.com/JuliaOpt/Mosek.jl/issues/92
 # and at least Mosek 8.0.0.41 for sosdemo6 to pass; see:
 # https://github.com/JuliaOpt/Mosek.jl/issues/98
-mos && push!(sdp_factories, with_optimizer(MosekTools.Mosek.Optimizer, QUIET=true))
+mos && push!(sdp_factories, optimizer_with_attributes(MosekTools.Mosek.Optimizer, "QUIET" => true))
 # Currently, need to create a file param.csdp to avoid printing, see https://github.com/JuliaOpt/CSDP.jl/issues/15
-csd && push!(sdp_factories, with_opt(CSDP, printlevel=0))
-iscsdp(factory) = occursin("CSDP", string(factory.constructor))
+# csd && push!(sdp_factories, optimizer_with_attributes(CSDP.Optimizer, "printlevel" => 0))
```
Comment (blegat) [inline test/solvers.jl]: Why is this commented ?

File: Project.toml [hunk 8]
```diff
@@ -17,8 +17,6 @@ SemialgebraicSets = "8e049039-38e8-557d-ae3a-bc521ccf6204"
...
 MathOptInterface = "~0.9.9"
 MultivariateBases = "0.1.1"
 MultivariateMoments = "0.3"
```
Comment (blegat) [inline Project.toml]: Bump PolyJuMP to v0.4

File: test/solvers.jl (hunk 7 above)
Comment (lkapelevich) [reply test/solvers.jl]: I should uncomment it. `Model(optimizer_with_attributes(CSDP.Optimizer, "printlevel" => 0.1))` would error. What's the best thing to do with solvers that haven't been updated for string parameters?

Comment (blegat) [reply test/solvers.jl]: Open an issue, I wasn't aware CSDP was not updated yet :)

**Comment by lkapelevich** [comment]
will require https://github.com/JuliaOpt/CSDP.jl/issues/53

**Comment by blegat** [comment]
CSDP should now work

**Comment by lkapelevich** [comment]
I don't think I can trigger the CI builds

**Pull request merged**